streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -223,22 +227,40 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
       thread.start()
     }
 
-    def stop() {
+    def stop(graceful: Boolean) {
       // Send the stop signal to all the receivers
       stopReceivers()
 
       // Wait for the Spark job that runs the receivers to be over
       // That is, for the receivers to quit gracefully.
+      val stopTimeout = conf.getLong(
+        "spark.streaming.gracefulStopTimeout",
+        10 * ssc.graph.batchDuration.milliseconds)
+
+
       thread.join(10000)
 
       // Check if all the receivers have been deregistered or not
-      if (!receiverInfo.isEmpty) {
-        logWarning("All of the receivers have not deregistered, " + receiverInfo)
+      def done = { receiverInfo.isEmpty && terminated }
+
+      if (graceful) {
+        TimeoutUtils.waitUntilDone(stopTimeout,() => done ) match {
+          case true =>
+            logInfo("All of the receivers have deregistered successfully and job terminated")
+          case false =>
+            logError("Timeout waiting for receivers to deregister")
+        }
       } else {
-        logInfo("All of the receivers have deregistered successfully")
+        if (!receiverInfo.isEmpty) {
```

Comment:
I think you might be able to remove a level of nesting and place this `if` alongside the `else` on the previous line.